python/qemu/ot/rom/image.py

```diff
@@ -249,54 +316,324 @@ def subst_perm_dec(cls, val, key, width, num_round):
+        inv = False
+        for mask in reversed(cls.ECC_39_32):
+            ecc <<= 1
+            parity = (data & mask).bit_count() & 1
+            ecc |= parity ^ int(inv)
+            inv = not inv
+        return (ecc << 32) | data
+
+    @classmethod
+    @lru_cache
+    def data_sp_enc(cls, val: int) -> int:
+        """Encode (scramble) data."""
+        return cls.subst_perm_enc(val, 0, cls.WORD_BITS,
+                                  cls.DATA_SUBST_PERM_ROUNDS)
+
     @classmethod
-    def data_sp_dec(cls, val):
+    @lru_cache
+    def data_sp_dec(cls, val: int) -> int:
+        """Decode (unscramble) data."""
         return cls.subst_perm_dec(val, 0, cls.WORD_BITS,
                                   cls.DATA_SUBST_PERM_ROUNDS)
 
+    def _load_hex(self, rfp: BinaryIO, size: Optional[int] = None) -> None:
+        words: list[int] = []  # 64-bit values
+        for lpos, line in enumerate(rfp.readlines(), start=1):
+            line = line.strip()
+            if len(line) != 10:
+                raise ValueError(f'Unsupported ROM HEX format at line {lpos}')
+            try:
+                words.append(int(line, 16))
+            except ValueError as exc:
+                raise ValueError(f'Invalid HEX data at line {lpos}: {exc}')
+        if size is not None and size != len(words) * 4:
+            raise ValueError('HEX content does not match ROM size')
+        self._handle_scrambled_data(words)
+
+    def _load_vmem(self, rfp: BinaryIO, size: Optional[int] = None) -> None:
+        words: list[int] = []  # 64-bit values
+        scrambled: Optional[bool] = None
+        next_addr = 0
+        for lpos, line in enumerate(rfp.readlines(), start=1):
+            line = line.strip()
+            if not line.startswith(b'@'):
+                continue
+            parts = line[1:].split(b' ')
+            address_str = parts[0]
+            data_str = parts[1:]
+            scrambled_data = all(len(d) == 10 for d in data_str)
+            plain_data = all(len(d) == 8 for d in data_str)
+            if not scrambled_data ^ plain_data:
+                raise ValueError(f'Unknown VMEM format @ {lpos}')
+            if scrambled is None:
+                scrambled = scrambled_data
+                self._log.info('Identified %s as %s VMEM format',
+                               rfp.name or '?',
+                               'scrambled' if scrambled else 'plain')
+            elif scrambled != scrambled_data:
+                raise ValueError(f'Incoherent VMEM format @ {lpos}')
+            try:
+                address = int(address_str, 16)
+                words.extend(int(d, 16) for d in data_str)
+            except (TypeError, ValueError) as exc:
+                raise ValueError(f'Invalid data in VMEM format @ '
+                                 f'{lpos}') from exc
+            if address != next_addr:
+                raise ValueError(f'Incoherent next address @ {lpos}')
+            next_addr += len(data_str)
```

Comment:
Maybe I'm missing something obvious (or just wrong), but it feels like this should be:

```suggestion
            next_addr += len(data_str) * (5 if scrambled_data else 4)
```
as I thought VMEM addresses were byte addresses, not word addresses.

Reply:
It seems to only count blocks/chunk, not addresses from what I observed in OT output files (weird format indeed...)

Reply:
Hmm, that's strange. When I've looked at the VMEM files we generate to give to Vivado they [definitely use byte addresses](https://github.com/AlexJones0/opentitan/blob/c174c90f9afa1d7bb265ce5fe007b80623b40f2f/hw/ip/rom_ctrl/util/gen_vivado_mem_image.py#L242C9-L242C13) but I'm not sure about the VMEMs used for e.g. Verilator that are produced by `srec_cat`.

Reply:
To be fair I never understood the point of a format encoding the number of chunks, since it is only redondant information, however, looking at OT output:

* scrambled file
```
head bazel-bin/sw/device/lib/testing/test_rom/test_rom_sim_verilator.39.scr.vmem
@00000000 53359F359C 2569E52275 3F96366673 041658761C 6201337727 6C0C3B9052
@00000006 5DF1CC4F20 0DA9BC7686 7ADB3794CA 02021B3D00 40233253D7 04DE2F873E
```

* plain file
```
head $BZLDJ/sw/device/tests/aes_smoketest_sim_verilator.32.vmem
/* http://srecord.sourceforge.net/ */
@00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000
@00000007 00000000 00000000 00000000 00000000 00000000 00000000 00000000
@0000000E 00000000 00000000 00000000 00000000 00000000 00000000 00000000
```

* OTP file
```
head img_rma.24.vmem
//
// OTP MEM file with 1024 x 24bit layout
@000000 000000 // VENDOR_TEST: SCRATCH
@000001 000000 // VENDOR_TEST: SCRATCH
@000002 000000 // VENDOR_TEST: SCRATCH
@000003 000000 // VENDOR_TEST: SCRATCH
```

None seems to encode the byte count, only the number of blocks which is already known :)

Reply:
So not even word addresses, only block count, whatever the size of a block

Reply:
I looked at the VMEMs generated by OpenTitan and you are correct, however it is definitely also the case that when generating Vivado updatemem-compatible files for FPGA (the script I linked) we do encode the byte counts, which I've checked:

```
head bazel-out/k8-fastbuild-ST-318f9db258e0/bin/hw/bitstream/universal/splice-rom.update.mem
@00000000 A015E74F07
@00000005 C349DC2D64
@0000000A 1A725E3D65
```

I did some searching in the history and found some relevant information - particularly notable is this commit: https://github.com/lowRISC/opentitan/commit/bb60413660225950ec27f772236c009cac923aed. The `srec_cat` tool, which we use for most VMEM conversions within Bazel, by default generates a linear addressing scheme (as you mention, redundant information?). Vivado's `updatemem` when mapping words to its BRAM architecture requires appropriate byte  addressing (see particularly the doc on the removed `addr4x.py` in that commit). Because I had previously looked at this script and not any `srec_cat` output I assumed all VMEMs were the same; this is apparently not the case.

So I think your current logic is correct for the majority of VMEMs created by OpenTitan, but would error (due to address mismatch) if you tried to input the VMEM used for rom/OTP splicing with Vivado. I think any way you decide to handle this is fine - you could just ignore it or add a special case. But it might be worth adding a note in the documentation that VMEMs are expected to have linear/chunk addressing.

Reply:
If I get it right from various sources, there is no unit definition for the address marker (as in `@address`): it is derivated from Verilog `$readmemh()` which counts in array index, i.e. there is no concept of unit (in terms of number of bytes) linked to it. It is a unitless value, which depends on the actual array. In other words, the producer and the consumer of this file format need to agree on a per file basis. Nice...

Reply:
Adding some comments to `_load_vmem` and `_print_vmem` functions.